CodeiumVS/QuickInfo/QuickInfoSource.cs

```diff
@@ -0,0 +1,150 @@
+using Microsoft.VisualStudio.Text.Adornments;
+using Microsoft.VisualStudio.Text.Editor;
+using Microsoft.VisualStudio.Text.Tagging;
+using Microsoft.VisualStudio.Utilities;
+using System.Collections.Generic;
+using System.ComponentModel.Composition;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace CodeiumVS.QuickInfo;
+
+
+internal sealed class CodeiumAsyncQuickInfoSource(ITextBuffer textBuffer) :
+    PropertyOwnerExtension<ITextBuffer, CodeiumAsyncQuickInfoSource>(textBuffer), IAsyncQuickInfoSource, IDisposable
+{
+    private ITextView _tagAggregatorTextView;
+    private ITagAggregator<IErrorTag>? _tagAggregator;
+
+    private static readonly ImageId _icon = KnownMonikers.StatusInformation.ToImageId();
+
+    private static string GetQuickInfoItemText(object content)
+    {
+        if (content is string str) return str;
+        if (content is ClassifiedTextRun textRun) return textRun.Text;
+
+        if (content is ContainerElement containter) 
+        {
+            string text = string.Empty;
+            foreach (var element in containter.Elements)
+                text += GetQuickInfoItemText(element);
+            return text;
+        }
+        
+        if (content is ClassifiedTextElement textElement) 
+        {
+            string text = string.Empty;
+            foreach (var element in textElement.Runs)
+                text += GetQuickInfoItemText(element);
+            return text;
+        }
+
+        return string.Empty;
+    }
+
+    public async Task<QuickInfoItem> GetQuickInfoItemAsync(IAsyncQuickInfoSession session,CancellationToken cancellationToken)
+    {
+        if (_disposed || session.TextView.TextBuffer != _owner) 
+            return null;
+
+        await GetTagAggregatorAsync(session.TextView);
+
+        Assumes.True(_tagAggregator != null,
+            "Codeium Quick Info Source couldn't create a tag aggregator for error tags");
+
+        // Put together the span over which tags are to be discovered.
+        // This will be the zero-length span at the trigger point of the session.
+        SnapshotPoint? subjectTriggerPoint = session.GetTriggerPoint(_owner.CurrentSnapshot);
+        if (!subjectTriggerPoint.HasValue)
+        {
+            Debug.Fail("The Codeium Quick Info Source is being called when it shouldn't be.");
+            return null;
+        }
+
+        ITextSnapshot currentSnapshot = subjectTriggerPoint.Value.Snapshot;
+        var querySpan = new SnapshotSpan(subjectTriggerPoint.Value, 0);
+
+        // Must be on main thread when dealing with tag aggregator
+        await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
+
+        // Ask for all of the error tags that intersect our query span.  We'll get back a list of mapping tag spans.
+        // The first of these is what we'll use for our quick info.
+        IEnumerable<IMappingTagSpan<IErrorTag>> tags = _tagAggregator.GetTags(querySpan);
```

Comment:
Note: All tasks related to the `ITagAggregator` must be on the main thread, it was hard tracking down this issue since the documentation doesn't mention anything about it.